scikeras/utils/transformers.py

```diff
@@ -0,0 +1,231 @@
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Union
+
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, OrdinalEncoder
+from tensorflow.python.keras.losses import is_categorical_crossentropy
+
+
+class Ensure2DTransformer(TransformerMixin, BaseEstimator):
+    """Transforms from 1D -> 2D and back.
+    """
+
+    def fit(self, X: np.ndarray) -> "Ensure2DTransformer":
+        self.should_transform_ = X.ndim == 1
+        return self
+
+    def transform(self, X: np.ndarray) -> np.ndarray:
+        if self.should_transform_:
+            if X.ndim != 1:
+                raise ValueError(
+                    "Expected a 1D array for `X`." f" Got a {X.ndim}D array instead."
+                )
+            X = X.reshape(-1, 1)
+        return X
+
+    def inverse_transform(self, X: np.ndarray) -> np.ndarray:
+        if self.should_transform_:
+            if X.ndim != 2:
+                raise ValueError(
+                    f"Expected a 2D array for `X`, but got a"
+                    f" {X.ndim}D array instead."
+                )
+            if X.shape[1] != 1:
+                raise ValueError(
+                    f"Expected `X.shape[1] == 1`"
+                    f" but got `X.shape[1] == {X.shape}` instead."
+                )
+            X = np.squeeze(X, axis=1)
+        return X
+
+
+class BaseKerasTransformer(TransformerMixin, BaseEstimator, ABC):
```

Comment:
The other methods have no implementation, only docstrings and type-hints. The types _are_ different from sklearn, since we expect/allow these transformers to return a dict of arrays or a list of arrays.

Reply:
Are we okay on this one? Or do you want to remove the docstrings/type-hints?

Reply:
I think the ability to use *any* Scikit-Learn transformer is more valuable than cluttering the documentation with `BaseKerasTransformer` and potentially confusing users. If I saw `BaseKerasTransformer` in the documentation as a user, I'd think "how should I make my Scikit-learn transformer fit into the framework?"

I think I'd remove `BaseKerasTransformer`; it doesn't do anything. `get_metadata` is only called if it exists, and it doesn't return anything in `BaseKerasTransformer`.